Loop/View Controllers/StatusTableViewController.swift

```diff
@@ -93,8 +95,11 @@ final class StatusTableViewController: LoopChartsTableViewController {
             notificationCenter.addObserver(forName: .HKUserPreferencesDidChange, object: deviceManager.glucoseStore.healthStore, queue: nil) {[weak self] _ in
                 DispatchQueue.main.async {
                     self?.log.debug("[reloadData] for HealthKit unit preference change")
-                    self?.preferredGlucoseUnit = self?.deviceManager.glucoseStore.preferredUnit
-                    self?.unitPreferencesDidChange(to: self?.preferredGlucoseUnit)
+                    if let glucoseUnit = self?.deviceManager.glucoseStore.preferredUnit {
```

Comment:
Hmm, if I understand correctly, this will be triggered on *any* HealthKit unit preferences change. Perhaps *only* the GlucoseStore itself should observe the `healthStore` for any `.HKUserPreferencesDidChange` and then check to see if it has actually changed, and then, if so, notify all observers on the GlucoseStore (with a new notification).

Also, I think I may see a timing problem with the `. HKUserPreferencesDidChange` notification from the health store. Specifically, the `glucoseStore.preferredUnit` is a computed property which is retrieved from the `HealthStoreUnitCache`. The `HealthStoreUnitCache` *also* observes to the `. HKUserPreferencesDidChange` notification in order to flush the unit cache when the preferences change. So, if the `HealthStoreUnitCache` gets the notification first, then the unit cache is flushed, and any other observers will pull the latest and correct unit from the store. However, if any other observer gets the notification first, it will query the `HealthStoreUnitCache` *and get the old, stale value from the cache*.

cc: @ps2 Thoughts?

Reply:
Also, this would affect not only here, but also at least two other observers of the `.HKUserPreferencesDidChange` notification (i.e. `CarbAbsorptionViewController`, `PredictionTableViewController`)

Reply:
Actually, I could see all of this functionality as part of the various data stores via `HealthKitSampleStore`.

Reply:
The docs say `Each HKHealthStore posts a HKUserPreferencesDidChangeNotification notification when the preferred unit for a HKQuantityType is changed by the user.` So it should only fire when the unit is changed. 

Thanks for pointing out the potential timing issue. I'll include an additional `glucoseStoreUnitChanged` notification that fires after `. HKUserPreferencesDidChange`.

Reply:
[Off today, but responding so you don't have to wait until tomorrow...]

Sorry, let me clarify. In the call to `notificationCenter.addObserver(forName: .HKUserPreferencesDidChange, object: deviceManager.glucoseStore.healthStore, queue: nil)` the `deviceManager.glucoseStore.healthStore` is an instance of `HKHealthStore`. As far as I know there is nothing that says "only notify me when the preferred unit for *just* blood glucose values changes". So, any time the user changes the preferred unit for *anything* changes(e.g. body mass, height, etc.), then it will go through this same logic even if the preferred unit for blood glucose values hasn't changed.

However, if the unit cache was the only one observing `HKUserPreferencesDidChange` on a `HKHealthStore` and it compared the current preferred unit with what was in its cache, then it could decide whether the cache needed to be updated or not and, if so, send out a separate notification about the blood glucose unit change (or any other unit that happened to be in its cache) to observers of the unit cache.